server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -670,6 +670,7 @@ public Searcher acquireSearcher(String source, SearcherScope scope) throws Engin
         }
         Releasable releasable = store::decRef;
         try {
+            assert assertSearcherIsWarmedUp(source, scope);
```

Comment:
Maybe also assert that searcher is warmed up when shard moves to `STARTED` state?

Reply:
Did you mean to expose this assertion to IndexShard?

Reply:
Not necessary, let us skip this. It would not catch the reset engine case anyway.